File: src/kernel/trap/fmt.zig
//! Trap Formatting Utilities.
//!
//! Lock-free formatting helpers for panic messages. These avoid std.fmt to work
//! in panic context where allocators and locks are unavailable.

/// Format a 64-bit value as hexadecimal (16 characters, no prefix).
pub fn formatHex(val: u64) [16]u8 {
    const hex_chars = "0123456789abcdef";
    var buf: [16]u8 = undefined;
    var v = val;
    var i: usize = 16;
    while (i > 0) {
        i -= 1;
        const nibble: usize = @truncate(v & 0xF);
        buf[i] = hex_chars[nibble];
        v >>= 4;
    }
    return buf;
}

/// Format a decimal number. Returns buffer and length of valid data.
pub fn formatDecimal(val: usize) struct { buf: [20]u8, len: usize } {
    var buf: [20]u8 = undefined;
    var v = val;
    var i: usize = buf.len;

    if (v == 0) {
        var result: [20]u8 = undefined;
        result[0] = '0';
        return .{ .buf = result, .len = 1 };
    }

    while (v > 0) : (i -= 1) {
        const digit: u8 = @truncate('0' + (v % 10));
        buf[i - 1] = digit;
        v /= 10;
    }

    const len = buf.len - i;
    var result: [20]u8 = undefined;
    for (0..len) |j| {
        result[j] = buf[i + j];
    }
    return .{ .buf = result, .len = len };
}

test "formatHex produces correct output" {
    const std = @import("std");
    try std.testing.expectEqualStrings("0000000000000000", &formatHex(0));
    try std.testing.expectEqualStrings("000000000000000f", &formatHex(0xF));
    try std.testing.expectEqualStrings("00000000deadbeef", &formatHex(0xDEADBEEF));
    try std.testing.expectEqualStrings("ffffffffffffffff", &formatHex(0xFFFFFFFFFFFFFFFF));
}

test "formatDecimal produces correct output" {
    const std = @import("std");
    const zero = formatDecimal(0);
    try std.testing.expectEqualStrings("0", zero.buf[0..zero.len]);

    const small = formatDecimal(42);
    try std.testing.expectEqualStrings("42", small.buf[0..small.len]);

    const large = formatDecimal(1234567890);
    try std.testing.expectEqualStrings("1234567890", large.buf[0..large.len]);
}

File: kernel/src/trap/fmt.rs
//! Trap formatting utilities.
//!
//! Panic and trap paths cannot allocate or take locks. These helpers format
//! numbers into fixed-size stack buffers so that output is always possible
//! even when the kernel is in distress.

const HEX_DIGITS: &[u8; 16] = b"0123456789abcdef";
const HEX_USIZE_DIGITS: usize = core::mem::size_of::<usize>() * 2;
const DECIMAL_USIZE_DIGITS: usize = 20;

/// Fixed-size buffer holding a decimal-formatted number.
pub struct Decimal {
    buf: [u8; DECIMAL_USIZE_DIGITS],
    len: usize,
}

impl Decimal {
    pub const fn as_bytes(&self) -> &[u8] {
        self.buf.split_at(self.len).0
    }

    pub const fn len(&self) -> usize {
        self.len
    }

    pub const fn is_empty(&self) -> bool {
        self.len == 0
    }
}

pub fn format_hex(value: u64) -> [u8; 16] {
    let mut buf = [0; 16];
    let mut shift: u32 = 60;
    let mut index = 0;

    while index < buf.len() {
        let nibble = ((value >> shift) & 0xf) as usize;
        buf[index] = HEX_DIGITS[nibble];
        index += 1;
        shift = shift.saturating_sub(4);
    }

    buf
}

pub fn format_hex_usize(value: usize) -> [u8; HEX_USIZE_DIGITS] {
    let mut buf = [0; HEX_USIZE_DIGITS];
    let mut index = 0;

    while index < buf.len() {
        let shift = (buf.len() - 1 - index) * 4;
        let nibble = (value >> shift) & 0xf;
        buf[index] = HEX_DIGITS[nibble];
        index += 1;
    }

    buf
}

pub fn format_decimal(value: usize) -> Decimal {
    let mut scratch = [0; DECIMAL_USIZE_DIGITS];
    let mut result = [0; DECIMAL_USIZE_DIGITS];
    let mut value = value;
    let mut index = scratch.len();

    if value == 0 {
        result[0] = b'0';
        return Decimal {
            buf: result,
            len: 1,
        };
    }

    while value > 0 {
        index -= 1;
        scratch[index] = b'0' + (value % 10) as u8;
        value /= 10;
    }

    let len = scratch.len() - index;
    result[..len].copy_from_slice(&scratch[index..]);

    Decimal { buf: result, len }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn formats_hex_output() {
        assert_eq!(&format_hex(0), b"0000000000000000");
        assert_eq!(&format_hex(0xf), b"000000000000000f");
        assert_eq!(&format_hex(0xdead_beef), b"00000000deadbeef");
        assert_eq!(&format_hex(u64::MAX), b"ffffffffffffffff");
    }

    #[test]
    fn formats_usize_hex_output() {
        assert_eq!(format_hex_usize(0)[0], b'0');
        assert_eq!(format_hex_usize(0)[HEX_USIZE_DIGITS - 1], b'0');
        assert_eq!(format_hex_usize(0xf)[HEX_USIZE_DIGITS - 1], b'f');
    }

    #[test]
    fn formats_decimal_output() {
        assert_eq!(format_decimal(0).as_bytes(), b"0");
        assert_eq!(format_decimal(42).as_bytes(), b"42");
        assert_eq!(format_decimal(1_234_567_890).as_bytes(), b"1234567890");
    }
}
